### PR TITLE
WIP: implement notifications

### PR DESCRIPTION
**Pull request by ylebre**
This PR is intended to add the notifications layer to the solid-nextcloud server. It took over the pubsub implementation from solid-crud and adds a webhook implementation as well.

### CODE REVIEW

**Review by Potherca — APPROVED** [review]

File: solid/lib/Controller/CalendarController.php [hunk 1]
```diff
@@ -118,15 +119,15 @@ public function handleRequest($userId, $path) {
...
 		$this->DPop = new DPop();
 
 		$request = $this->rawRequest;
 		$baseUrl = $this->getCalendarUrl($userId);		
 		$this->resourceServer->setBaseUrl($baseUrl);
 		$this->WAC->setBaseUrl($baseUrl);
-		$pubsub = getenv('PUBSUB_URL') ?: ("http://pubsub:8080/");
-		$this->resourceServer->setPubSubUrl($pubsub);
+		$notifications = new SolidNotifications();
+		$this->resourceServer->setNotifications($notification);
```
Comment (poef) [inline solid/lib/Controller/CalendarController.php:130]: Should this not be `setNotifications($notification**s**);` ?

File: solid/lib/Controller/ContactsController.php [hunk 2]
```diff
@@ -118,15 +119,15 @@ public function handleRequest($userId, $path) {
...
 		$this->DPop = new DPop();
 
 		$request = $this->rawRequest;
 		$baseUrl = $this->getContactsUrl($userId);		
 		$this->resourceServer->setBaseUrl($baseUrl);
 		$this->WAC->setBaseUrl($baseUrl);
-		$pubsub = getenv('PUBSUB_URL') ?: ("http://pubsub:8080/");
-		$this->resourceServer->setPubSubUrl($pubsub);
+		$notifications = new SolidNotifications();
+		$this->resourceServer->setNotifications($notification);
```
Comment (poef) [inline solid/lib/Controller/ContactsController.php:130]: Same problem here, `$notifications`

File: solid/lib/Controller/ProfileController.php [hunk 3]
```diff
@@ -135,15 +136,15 @@ public function handleRequest($userId, $path) {
...
 		$this->DPop = new DPop();
 
 		$request = $this->rawRequest;
 		$baseUrl = $this->getProfileUrl($userId);		
 		$this->resourceServer->setBaseUrl($baseUrl);
 		$this->WAC->setBaseUrl($baseUrl);
-		$pubsub = getenv('PUBSUB_URL') ?: ("http://pubsub:8080/");
-		$this->resourceServer->setPubSubUrl($pubsub);
+		$notifications = new SolidNotifications();
+		$this->resourceServer->setNotifications($notification);
```
Comment (poef) [inline solid/lib/Controller/ProfileController.php:148]: And here `$notifications`

File: solid/lib/Notifications/SolidWebHook.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                $postData = $this->createUpdatePayload($path, $type);
+                $opts = array('http' =>
+                    array(
+                        'method'  => 'POST',
+                        'header'  => 'Content-Type: application/ld+json',
+                        'content' => $postData
+                    )
+                );
+                $context  = stream_context_create($opts);
+                $result = file_get_contents($webhook, false, $context);
```
Comment (poef) [inline solid/lib/Notifications/SolidWebHook.php:26]: Where is `$webhook` coming from? Should that be `$url`?

File: solid/lib/Controller/ContactsController.php (hunk 2 above)
Comment (ylebre) [reply solid/lib/Controller/ContactsController.php]: yes, fixed in 943182c57ddac8e994b2dedb0097be27ab149624

File: solid/lib/Notifications/SolidWebHook.php (hunk 4 above)
Comment (ylebre) [reply solid/lib/Notifications/SolidWebHook.php]: yes - renamed them all to webhookUrl to be more clear.

**Review by Potherca — APPROVED** [review]

File: solid/lib/Service/SolidWebhookService.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		$webhook->setPath($path);
+		$webhook->setUrl($url);
+		$webhook->setExpiry($expiry);
+		return $this->mapper->insert($webhook);
+	}
+
+	public function delete($webId, $path) {
+		try {
+			$webhook = $this->mapper->find($webId, $path);
+			$this->mapper->delete($webhook);
```
Comment (Potherca) [inline solid/lib/Service/SolidWebhookService.php:60]: There does not seem to be a `delete` method in the SolidWebhookMapper class. Am I missing something, or is the method missing?

**Pull request merged**